### PR TITLE
Add support for ternary expression

### PR DESCRIPTION
**Pull request by satyamakgec**
### What was wrong?
#130 

### To-Do

- [x] Improve  `yul::mappers::expressions::exp_ternary` so it can return the valid `yul::Expression`.

[//]: # (Stay ahead of things, add list items here!)
- [ ] OPTIONAL: Update [Spec](https://github.com/ethereum/fe/blob/master/spec/index.md) if applicable
- [ ] Clean up commit history

### CODE REVIEW

File: compiler/src/yul/mappers/expressions.rs [hunk 1]
```diff
@@ -272,6 +272,27 @@ fn expr_attribute_self(
...
+        else_expr,
+    } = &exp.node
+    {
+        let yul_test_expr = expr(context, test)?;
+        let yul_if_expr = expr_comp_operation(context, if_expr)?;
+        let yul_else_exp = expr(context, else_expr)?;
+
+        // TODO: Need to find a way to return an expression that comprises all 3 of
+        // above. Dummy return to make it compile.
+        return Ok(literal_expression! {(1)});
```
Comment (cburgdorf) [inline compiler/src/yul/mappers/expressions.rs:291]: >My understanding is that we are missing the appropriate macro in the yulstur crate, As I can't find a macro or helper function that generates a valid expression from those 3 expressions, Or maybe I am missing something.

I'm going to answer this here because there's a chance that I'll end up giving you a wrong or misleading answer and so I want to page in @g-r-a-n-t to give him the chance to correct me if I'm talking nonsense :sweat_smile: 

So, what I'm thinking is that this would be best mapped to a `switch` statement in YUL. You can see this being used here:

https://github.com/ethereum/fe/blob/c2681b844bf441ee296356ea6ab41f86fbd1e125/compiler/src/yul/mappers/functions.rs#L107-L111

Now, the difference here would be the `switch` is a statement so it doesn't resolve to any value. So, what I'm thinking is that we would need to map this to a bigger code out to something like this:

Pseudo-YUL
```
switch test
case 0 { $result := <if-expression> }
case 1 { $result := <else-expression }
result
```

So, basically we would generate a switch *statement* that introduces a variable to store the result of each possible arm followed by an *expression* that simply resolves to whatever is behind `$result`.

Without getting my hand on the code to actually try it, this is what I think I would try first.

@g-r-a-n-t  Does this make sense?

File: semantics/src/traversal/expressions.rs [hunk 2]
```diff
@@ -323,6 +323,32 @@ fn expr_comp_operation(
...
+    context: Shared<Context>,
+    exp: &Spanned<fe::Expr>,
+) -> Result<ExpressionAttributes, SemanticError> {
+    if let fe::Expr::Ternary {
+        test,
+        if_expr,
+        else_expr,
+    } = &exp.node
+    {
+        let test_attributes = expr(Rc::clone(&scope), Rc::clone(&context), test);
```
Comment (cburgdorf) [inline semantics/src/traversal/expressions.rs:337]: One thing we need to check here is that the `test` is actually a boolean expression. Similar to

https://github.com/ethereum/fe/blob/c2681b844bf441ee296356ea6ab41f86fbd1e125/semantics/src/traversal/functions.rs#L232

Comment (satyamakgec) [reply semantics/src/traversal/expressions.rs:337]: ahh, the hardcoded boolean statement makes me confuse. No worries I will make that change.

FYI here we need to check the `if_attributes` as a boolean expression, not the `test` as the test consists of the successful return value. I think we need to work on the naming or make some re-arrangements if we are expecting to `test` as the boolean expression. 
Here is the declaration of the ternary expression https://github.com/ethereum/fe/blob/c2681b844bf441ee296356ea6ab41f86fbd1e125/parser/src/ast.rs#L196

File: compiler/src/yul/mappers/expressions.rs (hunk 1 above)
Comment (satyamakgec) [reply compiler/src/yul/mappers/expressions.rs:291]: > So, basically we would generate a switch statement that introduces a variable to store the result of each possible arm followed by an expression that simply resolves to whatever is behind $result.

The problem I saw is that we `switch!` is a statement I can't use to generate the expression so do you suggest using `yul::Switch` to generate the expression as you suggest above directly in the mappers instead of using the macros. 

Why do we need to store the result isn't we recursively generate the cases for it as per the given expressions.

File: semantics/src/traversal/expressions.rs (hunk 2 above)
Comment (cburgdorf) [reply semantics/src/traversal/expressions.rs:337]: Not sure I follow, let's break it down:

- `test`: This is the expression that is being tested e.g. `foo` in `1 if foo else 0`. The thing that we need to check for is that `foo` would would be an expression that evaluates to a `boolean` because non-booleans are illegal as a `test` in a ternary operator. E.g. we can not allow `foo` if it is of type `u256`.

- `if_expr`: The expression that is evaluated if `test` is `true`
- `else_expr`: The expression that is evaluated if `test` is false.

The thing that we need to check is that `if_expr` and `else_expr` evaluate to the same type. E.g. we can not allow one side to evaluate to a `address` if the other side evaluates to a `u256`

File: compiler/src/yul/mappers/expressions.rs (hunk 1 above)
Comment (cburgdorf) [reply compiler/src/yul/mappers/expressions.rs:291]: >The problem I saw is that we switch! is a statement I can't use to generate the expression

Correct. That's why I was thinking we could generate a switch that stores the result and then have that followed by an *expression* that just evaluates to whatever is behind that identifier. But as I'm looking over the various options, it's not immediately obvious to me how that could be done. @g-r-a-n-t can you help us out?

Comment (g-r-a-n-t) [reply compiler/src/yul/mappers/expressions.rs:291]: Ah, yeah this is tricky. In Yul we can't wrap statements within an expression like you can in Rust... so we fundamentally can't map this entirely on its own.

```rust
// an expression in Rust
{ 
  let a = 42;
  a
}
```

The work around for this should be to create a helper function that contains the terany statements and takes the relevant expressions as parameters. I'm thinking it should maybe look something like this:

```javascript
function teranary(test, if_expression, else_expression) -> result {
  switch test
  case 0 { result := <if-expression> }
  case 1 { result := <else-expression }
}
```

This ternary function would be included within the [standard](https://github.com/ethereum/fe/blob/c2681b844bf441ee296356ea6ab41f86fbd1e125/compiler/src/yul/runtime/functions.rs#L5-L20) set of functions made available during runtime, then when we map a ternary expression, it should looks something like this:

`1 if input > 5 else 0` -> `ternary(input > 5, 1, 0)` 

Hopefully this works and thank you for the commits. This was supposed to be an easy one :laughing:

Comment (g-r-a-n-t) [reply compiler/src/yul/mappers/expressions.rs:291]: I actually wouldn't really call it a work around. Languages that map to Yul are supposed to be human-readable, so it's good to keep expressions short.

File: semantics/src/traversal/expressions.rs (hunk 2 above)
Comment (satyamakgec) [reply semantics/src/traversal/expressions.rs:337]: I think this is the syntax we are trying to achieve as per Python grammar:
```
[on_true] if [expression] else [on_false]
```
So for our example

```
contract Foo:

    pub def bar(input: u256) -> u256:
        return 1 if input > 5 else 0
```

on_true = 1, expression = input > 5, on_false = 0

So by the parser, I got the Ternary expression values something like this - 

 `test` - Spanned { node: Num("1"), span: Span { start: 68, end: 69 } } which lead to `test_attribute` value is ExpressionAttributes { typ: Base(U256), location: Value }, So I don't think it will ever be boolean it will always be the evaluated value if `if_expression` is true.

`if_expr` - Spanned { node: CompOperation { left: Spanned { node: Name("input"), span: Span { start: 73, end: 78 } }, op: Spanned { node: Gt, span: Span { start: 79, end: 80 } }, right: Spanned { node: Num("5"), span: Span { start: 81, end: 82 } } }, span: Span { start: 73, end: 82 } } which lead to `if_expr_attribute` value is ExpressionAttributes { typ: Base(Bool), location: Value }. Which always be the boolean and it should be.

`else_expr` - Spanned { node: Num("0"), span: Span { start: 88, end: 89 } } which lead to `else_attribute` value is ExpressionAttributes { typ: Base(U256), location: Value } . This is expected result.

So what I mean above is as per your terminology `test` & `if_expr` value should be used interchangebly but as per the code, it is reverse or in other words, it is not as per your terminology you above mention. So what I suggest above either we change the Ternary declaration from 
```
Ternary {
        test: Box<Spanned<Expr<'a>>>,
        if_expr: Box<Spanned<Expr<'a>>>,
        else_expr: Box<Spanned<Expr<'a>>>,
    },  
to 
Ternary {
        if_expr: Box<Spanned<Expr<'a>>>,
        test: Box<Spanned<Expr<'a>>>,
        else_expr: Box<Spanned<Expr<'a>>>,
    },  
```
Or we can fix the parser.

BTW I already in place the check for the same type that is evaluated from the `if_expr` & `else_expr`.

Comment (cburgdorf) [reply semantics/src/traversal/expressions.rs:337]: >to 
```
Ternary {
        if_expr: Box<Spanned<Expr<'a>>>,
        test: Box<Spanned<Expr<'a>>>,
        else_expr: Box<Spanned<Expr<'a>>>,
    },
```

Ah, you mean changing the order of the enum properties to match the syntax of the language. We can definitely do that :+1:

File: compiler/src/yul/mappers/expressions.rs (hunk 1 above)
Comment (satyamakgec) [reply compiler/src/yul/mappers/expressions.rs]: Thanks, @g-r-a-n-t  & @cburgdorf it works.

File: semantics/src/traversal/expressions.rs (hunk 2 above)
Comment (satyamakgec) [reply semantics/src/traversal/expressions.rs]: Minor parser changes were needed to facilitate the ordering.

File: compiler/src/yul/mappers/expressions.rs [hunk 3]
```diff
@@ -272,6 +272,25 @@ fn expr_attribute_self(
...
+    context: &Context,
+    exp: &Spanned<fe::Expr>,
+) -> Result<yul::Expression, CompileError> {
+    if let fe::Expr::Ternary {
+        if_expr,
+        test,
+        else_expr,
+    } = &exp.node
+    {
+        let yul_test_expr = expr_comp_operation(context, test)?;
```
Comment (cburgdorf) [inline compiler/src/yul/mappers/expressions.rs:285]: The test needs to use `expr` as well. Otherwise cases like these would fail:

Boolean literal:

```
    pub def bar(input: u256) -> u256:
        return 1 if true else 0
```

Function call:

Boolean literal:

```
    pub def bar(input: u256) -> u256:
        return 1 if self.check(input) else 0
```

**Review by cburgdorf — APPROVED** [review]
Left one comment regarding the `expr_comp_operation` that needs to be changed to `expr` but otherwise looks good. Great work! :tada:

File: compiler/src/yul/mappers/expressions.rs (hunk 3 above)
Comment (satyamakgec) [reply compiler/src/yul/mappers/expressions.rs]: Thanks for the pointer. I fixed it.
